### PR TITLE
fix(Core/Conf): Fix AH searches with high number of auctions

### PR DESCRIPTION
**Pull request by MikaMauger**
The purpose of this PR is to fix a nasty bug when an auction search is a little bit too long (`GetMSTimeDiff(GameTime::GetGameTimeMS(), GetTimeMS()) >= 10ms`). This search is stopped, then restarted on next listing loop, but there is big chance this search will take more than 10ms again, looping forever, increasing CPU usage.

PS: this is my first PR on azerothcore, and my first time doing C++, so there will probably be a lot of things to improve.

## Changes Proposed:
- Decorelate auction listing from World update loop (there is a separate thread for this, no need to lock World update for auction searches)
- Setup a timeout for auction house search in worldserver.conf (default 1000ms)
- Discard the search if it lasts more than 1000ms (button will stay grey with no result for the player, but he can exit and launch UI again to get unstuck)
-  Some code cleaning

## Issues Addressed:

- N/A

## Tests Performed:

- Tested parallel searches with friends on 30k+ auctions, with almost instant responses

**ADMIN EDIT BY ACIDMANIFESTO NEW CONF WARNING**
Fo testers this adds a new conf to the world dist.:
```
#
#     AuctionHouse.SearchTimeout
#        Description: Time (in milliseconds) after which an auction house search is discarded.
#        Default:     1000 - (1 second)

AuctionHouse.SearchTimeout = 1000

```
**END ADMIN EDIT**

## How to Test the Changes:

1. Populate your auction house with a high amount of auctions (you can use mod-ah-bot for that)
2. Search items with name (with and without filters)
3. In worst case (could only reproduce it by lowering timeout or putting some sleep function), the search button will stay grey for the player and search is discarded from auction listing thread.

## Known Issues and TODO List:

There are a lot of stuff to do on the auction house side :
- some errors are not consistent with current action, some error codes are not even used
- code is a mess
- improve performances

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/AuctionHouse/AuctionHouseMgr.cpp [hunk 1]
```diff
@@ -752,20 +752,17 @@ bool AuctionHouseObject::BuildListAuctionItems(WorldPacket& data, Player* player
...
         int locdbc_idx = player->GetSession()->GetSessionDbcLocale();
 
-        for (AuctionEntryMap::const_iterator itr = AuctionsMap.begin(); itr != AuctionsMap.end(); ++itr)
+        for (auto &itr: AuctionsMap)
         {
-            if (!AsyncAuctionListingMgr::IsAuctionListingAllowed())                                                    // pussywizard: World::Update is waiting for us...
+            if ((itrcounter++) % 100 == 0) // check condition every 100 iterations
             {
-                if ((itrcounter++) % 100 == 0) // check condition every 100 iterations
+                if (GetMSTimeDiff(GameTime::GetGameTimeMS(), GetTimeMS()) >= searchTimeout) // pussywizard: stop immediately if diff is high or waiting too long
```
Comment (acidmanifesto) [inline src/server/game/AuctionHouse/AuctionHouseMgr.cpp:759]: everytime i see that name i keep thinking that guy must of had some serious confidence.

**Review by acidmanifesto — APPROVED** [review]

File: src/server/game/AuctionHouse/AuctionHouseMgr.cpp (hunk 1 above)
Comment (MikaMauger) [reply src/server/game/AuctionHouse/AuctionHouseMgr.cpp:759]: Ah ah indeed

**Comment by acidmanifesto** [comment]
@azerothcore/testers if able. this will be a performance improvement.

**Comment by Maelthyrr** [comment]
Hi, could you upstream your PR ? I've some merge conflicts for tests.

**Comment by MikaMauger** [comment]
> Hi, could you upstream your PR ? I've some merge conflicts for tests.

Done.

**Comment by Maelthyrr** [comment]
Hi, thanks for the PR :) I tested in solo with two char by selling munitions (around 10k). I see any issues with selling, buying, researching...

**Comment by acidmanifesto** [comment]
> Hi, thanks for the PR :) I tested in solo with two char by selling munitions (around 10k). I see any issues with selling, buying, researching... 

You do or don't see any issues?

**Comment by Maelthyrr** [comment]
> > Hi, thanks for the PR :) I tested in solo with two char by selling munitions (around 10k). I see any issues with selling, buying, researching...
> 
> You do or don't see any issues?

Fix sorry ^^. Everything looks fine in local.

**Comment by acidmanifesto** [comment]
👍

**Comment by MikaMauger** [comment]
I think we can close this PR in favor of https://github.com/azerothcore/azerothcore-wotlk/pull/15334

**Review by Kitzunu — APPROVED** [review]
Approving as previous staff approved and previous tests showed no issues with it

**Pull request merged**